File: python_backend/mistake_detector.py
from __future__ import annotations

import datetime as _dt
import uuid
from typing import Any, Dict, List

from .context_builder import ContextBuilder
from .llm import LLMClient, PROMPTS, format_sample_data, format_sheet_context


class MistakeDetector:
  """
  Port of the TypeScript MistakeDetector. Uses rule-based checks plus an LLM
  to detect potential issues in a Google Sheet.
  """

  def __init__(self, context_builder: ContextBuilder, llm_client: LLMClient) -> None:
    self.context_builder = context_builder
    self.llm_client = llm_client

  def detect_issues(
    self,
    spreadsheet_id: str,
    sheet_title: str,
    config: Dict[str, Any],
  ) -> Dict[str, Any]:
    context = self.context_builder.build_context(spreadsheet_id, sheet_title)
    issues: List[Dict[str, Any]] = []

    if config.get("enableRuleBased"):
      issues.extend(self._run_rule_based_checks(context, config))

    if config.get("enableLLMBased"):
      issues.extend(self._run_llm_based_checks(context, config))

    filtered = self._filter_by_severity(issues, config.get("minSeverity", "info"))
    max_issues = config.get("maxIssues")
    final_issues = filtered[:max_issues] if max_issues else filtered

    return {
      "spreadsheetId": spreadsheet_id,
      "sheetTitle": sheet_title,
      "issues": final_issues,
      "summary": self._generate_summary(final_issues),
      "scanTimestamp": _dt.datetime.utcnow().isoformat() + "Z",
    }

  # --- rule-based checks ---

  def _run_rule_based_checks(self, context: Dict[str, Any], config: Dict[str, Any]) -> List[Dict[str, Any]]:
    issues: List[Dict[str, Any]] = []
    categories: List[str] = config.get("categoriesToCheck") or []

    if "formula_error" in categories:
      issues.extend(self._check_formula_errors(context))
    if "inconsistent_formula" in categories:
      issues.extend(self._check_inconsistent_formulas(context))
    if "type_mismatch" in categories:
      issues.extend(self._check_type_mismatches(context))
    if "missing_value" in categories:
      issues.extend(self._check_missing_values(context))
    if "duplicate_key" in categories:
      issues.extend(self._check_duplicate_keys(context))

    return issues

  def _check_formula_errors(self, context: Dict[str, Any]) -> List[Dict[str, Any]]:
    summary = context.get("summary") or {}
    error_cells = summary.get("errorCells", 0)
    if error_cells <= 0:
      return []

    sheet_meta = context.get("sheetMetadata") or {}
    row_count = sheet_meta.get("rowCount", 0)

    return [
      {
        "id": str(uuid.uuid4()),
        "category": "formula_error",
        "severity": "high",
        "title": "Formula errors detected",
        "description": f"Found {error_cells} cells with formula errors (e.g., #REF!, #VALUE!, #DIV/0!).",
        "ranges": [
          {
            "a1Notation": f"1:{row_count}",
            "description": f"{error_cells} cell{'s' if error_cells > 1 else ''} with errors",
            "cellCount": error_cells,
          }
        ],
        "affectedCells": error_cells,
        "detectedBy": "rule",
        "suggestedFix": "Review and fix formula references and calculations.",
        "autoFixable": False,
      }
    ]

  def _check_inconsistent_formulas(self, context: Dict[str, Any]) -> List[Dict[str, Any]]:
    # Placeholder: requires more detailed data analysis
    return []

  def _check_type_mismatches(self, context: Dict[str, Any]) -> List[Dict[str, Any]]:
    issues: List[Dict[str, Any]] = []
    table_regions = context.get("tableRegions") or []

    for region in table_regions:
      for col in region.get("columns") or []:
        if col.get("type") == "mixed":
          col_index = col.get("index", 0)
          col_letter = self._column_to_letter(col_index + 1)
          issues.append(
            {
              "id": str(uuid.uuid4()),
              "category": "type_mismatch",
              "severity": "medium",
              "title": f'Mixed data types in column "{col.get("name")}"',
              "description": "Column contains multiple data types. This may indicate data quality issues.",
              "ranges": [
                {
                  "a1Notation": f"{col_letter}:{col_letter}",
                  "description": f"Column {col_letter} ({col.get('name')})",
                }
              ],
              "detectedBy": "rule",
              "suggestedFix": "Standardize data types in this column.",
              "autoFixable": False,
            }
          )

    return issues

  def _check_missing_values(self, context: Dict[str, Any]) -> List[Dict[str, Any]]:
    issues: List[Dict[str, Any]] = []
    table_regions = context.get("tableRegions") or []

    for region in table_regions:
      for col in region.get("columns") or []:
        # Mirror TS heuristic: first 3 columns should not be nullable
        if col.get("nullable") and col.get("index", 0) < 3:
          col_index = col.get("index", 0)
          col_letter = self._column_to_letter(col_index + 1)
          issues.append(
            {
              "id": str(uuid.uuid4()),
              "category": "missing_value",
              "severity": "low",
              "title": f'Missing values in key column "{col.get("name")}"',
              "description": "Key column contains empty cells which may indicate incomplete data.",
              "ranges": [
                {
                  "a1Notation": f"{col_letter}:{col_letter}",
                  "description": f"Column {col_letter} ({col.get('name')})",
                }
              ],
              "detectedBy": "rule",
              "suggestedFix": "Fill in missing values or mark as N/A.",
              "autoFixable": False,
            }
          )

    return issues

  def _check_duplicate_keys(self, context: Dict[str, Any]) -> List[Dict[str, Any]]:
    # Placeholder: would need full data scan to detect duplicates
    return []

  # --- LLM-based checks ---

  def _run_llm_based_checks(self, context: Dict[str, Any], config: Dict[str, Any]) -> List[Dict[str, Any]]:
    try:
      context_str = format_sheet_context(context)
      sample_ranges = context.get("sampleData") or []
      if sample_ranges:
        sample_data_str = format_sample_data(sample_ranges[0].get("values") or [])
      else:
        sample_data_str = "No sample data"

      user_prompt = PROMPTS.MISTAKE_DETECTION.user(context_str, sample_data_str)

      response = self.llm_client.chat_json(
        [
          {"role": "system", "content": PROMPTS.MISTAKE_DETECTION.system},
          {"role": "user", "content": user_prompt},
        ],
        overrides={"temperature": 0.3},
      )

      issues: List[Dict[str, Any]] = []
      for item in response:
        ranges = item.get("ranges")
        if not ranges and item.get("location"):
          ranges = self._parse_location_to_ranges(item["location"])

        issues.append(
          {
            "id": str(uuid.uuid4()),
            "category": item.get("category"),
            "severity": item.get("severity"),
            "title": item.get("title"),
            "description": item.get("description"),
            "ranges": ranges or [],
            "detectedBy": "llm",
            "suggestedFix": item.get("suggestedFix"),
            "autoFixable": False,
            "confidence": item.get("confidence", 0.8),
          }
        )

      return issues
    except Exception:
      # In case of LLM failure, fall back silently (log could be added)
      return []

  @staticmethod
  def _parse_location_to_ranges(location: str) -> List[Dict[str, Any]]:
    # Simple fallback for legacy location string
    return [
      {
        "a1Notation": location,
        "description": location,
      }
    ]

  # --- helpers ---

  @staticmethod
  def _filter_by_severity(issues: List[Dict[str, Any]], min_severity: str) -> List[Dict[str, Any]]:
    severity_order = ["critical", "high", "medium", "low", "info"]
    try:
      min_index = severity_order.index(min_severity)
    except ValueError:
      min_index = len(severity_order) - 1

    filtered: List[Dict[str, Any]] = []
    for issue in issues:
      sev = issue.get("severity", "info")
      try:
        issue_index = severity_order.index(sev)
      except ValueError:
        issue_index = len(severity_order) - 1
      if issue_index <= min_index:
        filtered.append(issue)
    return filtered

  @staticmethod
  def _generate_summary(issues: List[Dict[str, Any]]) -> Dict[str, Any]:
    severities = ["critical", "high", "medium", "low", "info"]
    categories = [
      "formula_error",
      "inconsistent_formula",
      "type_mismatch",
      "outlier",
      "missing_value",
      "broken_reference",
      "duplicate_key",
      "constraint_violation",
      "semantic_anomaly",
      "logical_inconsistency",
      "suspicious_pattern",
    ]

    by_severity = {s: 0 for s in severities}
    by_category = {c: 0 for c in categories}
    auto_fixable_count = 0

    for issue in issues:
      sev = issue.get("severity")
      cat = issue.get("category")
      if sev in by_severity:
        by_severity[sev] += 1
      if cat in by_category:
        by_category[cat] += 1
      if issue.get("autoFixable"):
        auto_fixable_count += 1

    return {
      "totalIssues": len(issues),
      "bySeverity": by_severity,
      "byCategory": by_category,
      "autoFixableCount": auto_fixable_count,
    }

  @staticmethod
  def _column_to_letter(column: int) -> str:
    letter = ""
    while column > 0:
      remainder = (column - 1) % 26
      letter = chr(65 + remainder) + letter
      column = (column - 1) // 26
    return letter

File: lib/agents/mistake-detector.ts
// Mistake detection engine combining rule-based and LLM-based analysis

import { LLMClient } from '../llm/client';
import { ContextBuilder } from '../sheets/context-builder';
import {
  SheetIssue,
  IssueDetectionResult,
  IssueDetectionConfig,
  ErrorSeverity,
  ErrorCategory,
} from '@/types/errors';
import { SheetContext, CellValue } from '@/types/sheets';
import { PROMPTS, formatSheetContext, formatSampleData } from '../llm/prompts';
import { v4 as uuidv4 } from 'uuid';

export class MistakeDetector {
  constructor(
    private contextBuilder: ContextBuilder,
    private llmClient: LLMClient
  ) {}

  /**
   * Detect issues in a sheet
   */
  async detectIssues(
    spreadsheetId: string,
    sheetTitle: string,
    config: IssueDetectionConfig
  ): Promise<IssueDetectionResult> {
    const context = await this.contextBuilder.buildContext(spreadsheetId, sheetTitle);
    const issues: SheetIssue[] = [];

    // Run rule-based checks
    if (config.enableRuleBased) {
      const ruleBasedIssues = await this.runRuleBasedChecks(context, config);
      issues.push(...ruleBasedIssues);
    }

    // Run LLM-based checks
    if (config.enableLLMBased) {
      const llmIssues = await this.runLLMBasedChecks(context, config);
      issues.push(...llmIssues);
    }

    // Filter by severity
    const filteredIssues = this.filterBySeverity(issues, config.minSeverity);

    // Limit number of issues if specified
    const finalIssues = config.maxIssues
      ? filteredIssues.slice(0, config.maxIssues)
      : filteredIssues;

    return {
      spreadsheetId,
      sheetTitle,
      issues: finalIssues,
      summary: this.generateSummary(finalIssues),
      scanTimestamp: new Date().toISOString(),
    };
  }

  /**
   * Run rule-based checks
   */
  private async runRuleBasedChecks(
    context: SheetContext,
    config: IssueDetectionConfig
  ): Promise<SheetIssue[]> {
    const issues: SheetIssue[] = [];

    // Check for formula errors
    if (config.categoriesToCheck.includes('formula_error')) {
      issues.push(...this.checkFormulaErrors(context));
    }

    // Check for inconsistent formulas
    if (config.categoriesToCheck.includes('inconsistent_formula')) {
      issues.push(...this.checkInconsistentFormulas(context));
    }

    // Check for type mismatches
    if (config.categoriesToCheck.includes('type_mismatch')) {
      issues.push(...this.checkTypeMismatches(context));
    }

    // Check for missing values in key columns
    if (config.categoriesToCheck.includes('missing_value')) {
      issues.push(...this.checkMissingValues(context));
    }

    // Check for duplicate keys
    if (config.categoriesToCheck.includes('duplicate_key')) {
      issues.push(...this.checkDuplicateKeys(context));
    }

    return issues;
  }

  /**
   * Check for formula errors
   */
  private checkFormulaErrors(context: SheetContext): SheetIssue[] {
    const issues: SheetIssue[] = [];

    if (context.summary.errorCells > 0) {
      // In a real implementation, we'd iterate through cells to find exact locations
      issues.push({
        id: uuidv4(),
        category: 'formula_error',
        severity: 'high',
        title: 'Formula errors detected',
        description: `Found ${context.summary.errorCells} cells with formula errors (e.g., #REF!, #VALUE!, #DIV/0!)`,
        location: {
          sheet: context.sheetMetadata.title,
          row: 0,
          col: 0,
          a1Notation: context.sheetMetadata.title,
        },
        affectedCells: context.summary.errorCells,
        detectedBy: 'rule',
        suggestedFix: 'Review and fix formula references and calculations',
        autoFixable: false,
      });
    }

    return issues;
  }

  /**
   * Check for inconsistent formulas in columns
   */
  private checkInconsistentFormulas(context: SheetContext): SheetIssue[] {
    const issues: SheetIssue[] = [];

    // This would require analyzing formula patterns in columns
    // For now, return empty array (implement with actual data access)

    return issues;
  }

  /**
   * Check for type mismatches in columns
   */
  private checkTypeMismatches(context: SheetContext): SheetIssue[] {
    const issues: SheetIssue[] = [];

    context.tableRegions.forEach(region => {
      region.columns.forEach(col => {
        if (col.type === 'mixed') {
          issues.push({
            id: uuidv4(),
            category: 'type_mismatch',
            severity: 'medium',
            title: `Mixed data types in column "${col.name}"`,
            description: `Column contains multiple data types. This may indicate data quality issues.`,
            location: {
              sheet: context.sheetMetadata.title,
              row: 0,
              col: col.index,
              a1Notation: `${context.sheetMetadata.title}!${this.columnToLetter(col.index + 1)}:${this.columnToLetter(col.index + 1)}`,
            },
            detectedBy: 'rule',
            suggestedFix: 'Standardize data types in this column',
            autoFixable: false,
          });
        }
      });
    });

    return issues;
  }

  /**
   * Check for missing values in non-nullable columns
   */
  private checkMissingValues(context: SheetContext): SheetIssue[] {
    const issues: SheetIssue[] = [];

    context.tableRegions.forEach(region => {
      region.columns.forEach(col => {
        if (col.nullable && col.index < 3) {
          // Assume first 3 columns should not be nullable (ID, name, etc.)
          issues.push({
            id: uuidv4(),
            category: 'missing_value',
            severity: 'low',
            title: `Missing values in key column "${col.name}"`,
            description: `Key column contains empty cells which may indicate incomplete data.`,
            location: {
              sheet: context.sheetMetadata.title,
              row: 0,
              col: col.index,
              a1Notation: `${context.sheetMetadata.title}!${this.columnToLetter(col.index + 1)}:${this.columnToLetter(col.index + 1)}`,
            },
            detectedBy: 'rule',
            suggestedFix: 'Fill in missing values or mark as N/A',
            autoFixable: false,
          });
        }
      });
    });

    return issues;
  }

  /**
   * Check for duplicate keys
   */
  private checkDuplicateKeys(context: SheetContext): SheetIssue[] {
    const issues: SheetIssue[] = [];

    // Would need to analyze actual data to detect duplicates
    // For now, return empty array

    return issues;
  }

  /**
   * Run LLM-based checks for semantic and logical issues
   */
  private async runLLMBasedChecks(
    context: SheetContext,
    config: IssueDetectionConfig
  ): Promise<SheetIssue[]> {
    try {
      const contextStr = formatSheetContext(context);
      const sampleDataStr = context.sampleData
        ? formatSampleData(context.sampleData[0]?.values || [])
        : 'No sample data';

      const userPrompt = PROMPTS.MISTAKE_DETECTION.user(contextStr, sampleDataStr);

      const response = await this.llmClient.chatJSON<any[]>(
        [
          { role: 'system', content: PROMPTS.MISTAKE_DETECTION.system },
          { role: 'user', content: userPrompt },
        ],
        { temperature: 0.3 } // Lower temperature for more consistent analysis
      );

      // Map LLM response to SheetIssue format
      const issues: SheetIssue[] = response.map((item: any) => ({
        id: uuidv4(),
        category: item.category as ErrorCategory,
        severity: item.severity as ErrorSeverity,
        title: item.title,
        description: item.description,
        location: this.parseLocation(item.location, context.sheetMetadata.title),
        detectedBy: 'llm',
        suggestedFix: item.suggestedFix,
        autoFixable: false,
        confidence: item.confidence || 0.8,
      }));

      return issues;
    } catch (error) {
      console.error('LLM-based detection failed:', error);
      return [];
    }
  }

  /**
   * Parse location string to CellLocation
   */
  private parseLocation(location: string, sheetTitle: string): any {
    // Simple A1 notation parser
    return {
      sheet: sheetTitle,
      row: 0,
      col: 0,
      a1Notation: location,
    };
  }

  /**
   * Filter issues by severity
   */
  private filterBySeverity(issues: SheetIssue[], minSeverity: ErrorSeverity): SheetIssue[] {
    const severityOrder: ErrorSeverity[] = ['critical', 'high', 'medium', 'low', 'info'];
    const minIndex = severityOrder.indexOf(minSeverity);

    return issues.filter(issue => {
      const issueIndex = severityOrder.indexOf(issue.severity);
      return issueIndex <= minIndex;
    });
  }

  /**
   * Generate summary statistics
   */
  private generateSummary(issues: SheetIssue[]) {
    const bySeverity: Record<ErrorSeverity, number> = {
      critical: 0,
      high: 0,
      medium: 0,
      low: 0,
      info: 0,
    };

    const byCategory: Record<ErrorCategory, number> = {
      formula_error: 0,
      inconsistent_formula: 0,
      type_mismatch: 0,
      outlier: 0,
      missing_value: 0,
      broken_reference: 0,
      duplicate_key: 0,
      constraint_violation: 0,
      semantic_anomaly: 0,
      logical_inconsistency: 0,
      suspicious_pattern: 0,
    };

    let autoFixableCount = 0;

    issues.forEach(issue => {
      bySeverity[issue.severity]++;
      byCategory[issue.category]++;
      if (issue.autoFixable) autoFixableCount++;
    });

    return {
      totalIssues: issues.length,
      bySeverity,
      byCategory,
      autoFixableCount,
    };
  }

  /**
   * Convert column number to letter
   */
  private columnToLetter(column: number): string {
    let letter = '';
    while (column > 0) {
      const remainder = (column - 1) % 26;
      letter = String.fromCharCode(65 + remainder) + letter;
      column = Math.floor((column - 1) / 26);
    }
    return letter;
  }
}
